### ISSUE TITLE
bug(ctm): ctm_tensor_c4v converges to a different fixed point than ctm_tensor on ~1/3 of random D=3 states with TRIVIAL virtual charges

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#833]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

`ctm_tensor_c4v` converges to a **different fixed point** than `ctm_tensor` on
roughly a third of random D=3 C4v-symmetric states with **all-zero virtual
charges** — i.e. with no block-sparse structure at all, so this is reachable
with no U(1) charges in play. Distinct from #762 (block-sparse RDM pairing) and
#763 (per-sector χ allocation), both of which need non-trivial virtual charges.

D=2 is clean. This is the residue that #760/#761 did not cure.

## Evidence

Fixtures: random tensor → `symmetrize_c4v` → all-zero virtual and physical
charges (so `SymmetricTensor` structure is absent; these are effectively dense).
Both paths at `max_iter=1000, conv_tol=1e-14`, `projector_method="eigh"`,
Heisenberg gate, χ=12.

| D | seeds tested | disagree | gap |
|---|---|---|---|
| 2 | 7, 11, 13, 17, 23, 29 | 1/6 (seed 17) | 1.5e-08 |
| 3 | 7, 11, 13, 17, 23, 29 | **2/6 (seeds 7, 17)** | **7.6e-04, 1.7e-03** |

The gap does not close with χ, and the C4v side is non-monotone while the
general side converges:

```
D=3 seed=7, tight stop                    D=3 seed=11 (control)
 chi   E_general        E_c4v     rel      chi   E_general        E_c4v     rel
   8  0.437203010  0.436969082  5.4e-04      8  0.484516368  0.484516368  0.0e+00
  12  0.437221398  0.436888164  7.6e-04     12  0.484503911  0.484503911  6.9e-16
  16  0.437226331  0.436852271  8.6e-04     16  0.484499088  0.484499088  0.0e+00
  24  0.437228093  0.436873209  8.1e-04     24  0.484499096  0.484499096  0.0e+00
```

`E_general` is monotone (0.437203 → 0.437221 → 0.437226 → 0.437228); `E_c4v`
wanders (0.436969 → 0.436888 → 0.436852 → 0.436873). That asymmetry is the
reason to suspect the C4v side rather than call it a truncation difference.

For D=2 the same sweep gives ≤6.5e-15 at every χ on every seed except 17.

## It is not non-convergence

Tightening the stopping criterion at D=3 seed 7, χ=12:

```
 max_iter  conv_tol     E_general            E_c4v          rel gap
      100     1e-09  0.436889710026  0.436888152131        3.566e-06
      300     1e-12  0.437221398083  0.436888163535        7.622e-04
     1000     1e-14  0.437221398086  0.436888163542        7.622e-04
```

Both sides are individually stable to 10 digits between the last two rows, and
the gap plateaus. **Note the trap in row 1:** at loose tolerance the gap reads
3.6e-06 and looks acceptable, only because the *general* side has not converged
yet. Any comparison of two CTMs at `conv_tol=1e-9` is uninformative.

## Spectral degeneracy is REFUTED as the cause

The obvious hypothesis — a near-degenerate pair at the truncation edge leaving
the `eigh` eigenbasis underdetermined — does not survive the data. Corner
spectrum of the general environment, χ=12:

| D | seed | rel gap | edge spacing | spectrum[:4] | |
|---|---|---|---|---|---|
| 2 | 17 | 1.5e-08 | **3.0e-04** | `[1, 0.9995, 0.458, 0.4578]` | ← degenerate, and tiny gap |
| 3 | 7 | **7.6e-04** | 1.1e-01 | `[1, 0.322, 0.274, 0.155]` | ← wide spacing |
| 3 | 17 | **1.7e-03** | 2.3e-01 | `[1, 0.256, 0.220, 0.143]` | ← wide spacing |
| 3 | 13 | 4.2e-15 | 8.9e-03 | `[1, 0.653, 0.314, 0.278]` | ← tighter, yet exact |
| 3 | 23 | 2.3e-16 | 1.7e-01 | `[1, 0.0994, 0.0991, 0.0905]` | ← degenerate pair, yet exact |

The D=2 outlier does fit the degeneracy story and is five orders smaller. The
two D=3 cases have *wide* edge spacings, while seeds with tighter spacings and
outright degenerate pairs agree to machine precision. Please don't re-run this
hypothesis.

## Not established

- **Whether a physical (optimized) D=3 state hits it.** Only random
  C4v-projected fixtures were tested. This is the check that decides whether it
  matters in practice. Bosonic simple update is broken at D≥3 (see #760), so
  producing one means `optimize_gs_ad` *without* `gs_c4v`, then projecting.
- Any shared mechanism with #762's dense arm. Plausible, unproven.
- Whether the D=2 seed-17 case (1.5e-08, genuinely degenerate) is the same
  defect or ordinary basis-selection noise.

## Why no test catches it

`tests/test_ctm_tensor_c4v.py::test_energy_matches_general_ctm_on_an_entangled_state`
is the guard for exactly this class (added for #760). It pins **one seed, D=2,
χ=8**. Every failure above is D=3 and seed-dependent, so it passes throughout.
A multi-seed D=3 version is the missing coverage.

## Reproduction

```python
import numpy as np, jax.numpy as jnp
from tenax.core.index import TensorIndex, FlowDirection
from tenax.core.symmetry import U1Symmetry
from tenax.core.tensor import DenseTensor
from tenax.algorithms._ctm_tensor_convergence import ctm_tensor
from tenax.algorithms._ctm_tensor_c4v import ctm_tensor_c4v
from tenax.algorithms._ctm_tensor_energy import compute_energy_ctm_tensor
from tenax.algorithms.ipeps import symmetrize_c4v

def state(D, seed, d=2):
    rng = np.random.RandomState(seed)
    data = symmetrize_c4v(jnp.array(rng.standard_normal((D, D, D, D, d))))
    data = data / (jnp.linalg.norm(data) + 1e-10)
    sym, zc, zp = U1Symmetry(), np.zeros(D, np.int32), np.zeros(d, np.int32)
    idx = (
        TensorIndex.from_charges(sym, zc.copy(), FlowDirection.OUT, label="u"),
        TensorIndex.from_charges(sym, zc.copy(), FlowDirection.IN, label="d"),
        TensorIndex.from_charges(sym, zc.copy(), FlowDirection.OUT, label="l"),
        TensorIndex.from_charges(sym, zc.copy(), FlowDirection.IN, label="r"),
        TensorIndex.from_charges(sym, zp.copy(), FlowDirection.IN, label="phys"),
    )
    return DenseTensor(data, idx)

Sz = 0.5 * jnp.array([[1.0, 0.0], [0.0, -1.0]])
Sp = jnp.array([[0.0, 1.0], [0.0, 0.0]]); Sm = jnp.array([[0.0, 0.0], [1.0, 0.0]])
H = (jnp.kron(Sz, Sz) + 0.5*jnp.kron(Sp, Sm) + 0.5*jnp.kron(Sm, Sp)).reshape(2, 2, 2, 2)

A = state(3, 7)
KW = dict(max_iter=1000, conv_tol=1e-14)
env, _ = ctm_tensor(A, chi=12, projector_method="eigh", **KW)
E_gen = float(compute_energy_ctm_tensor(A, env, H, d=2))
E_c4v = float(compute_energy_ctm_tensor(A, ctm_tensor_c4v(A, chi=12, **KW), H, d=2))
print(E_gen, E_c4v, abs(E_c4v - E_gen) / abs(E_gen))   # -> 7.622e-04
```

Roughly 2 min on CPU. Seeds 11/13/23/29 give ~1e-16 through the same code.

Refs #760, #761, #762, #763.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Progress on the first "Not established" item — *whether a physical (optimized) state hits it*. This is the **D=2 control** only; the D=3 arm is running and I'll post it separately whichever way it falls.

**D=2 physical states are clean, and more cleanly than the random fixtures were.**

## Recipe

Per the issue: `optimize_gs_ad` with `gs_c4v=False`, so `ctm_tensor_c4v` is nowhere in the optimization loop and the comparison isn't circular. Changes from a first attempt that failed (below): L-BFGS instead of Adam, 200 steps, and a C4v-symmetric starting point (`symmetrize_c4v` of a random tensor, deviation `1.4e-16`). Optimized at χ=12; compared at χ=8/12/16/24 with `max_iter=1000, conv_tol=1e-14`, `projector_method="eigh"`.

Two gates before believing any comparison:

```
G1  E_opt must reach the known D=2 value −0.6599411913   (else not converged)
G2  ‖P(A)−A‖/‖A‖ must be small                           (else the state isn't C4v
                                                          and the C4v CTM doesn't apply)
```

## Result

```
E_opt = −0.659731739524   (200 L-BFGS steps, χ_opt=12, 800 s on one A100)
G1: |E_opt − ref|/|ref| = 3.174e-04                        → PASS
G2: C4v deviation of the optimized state = 1.844e-01       → FAIL
```

```
--- D=2 optimized, as-is ---            --- D=2 optimized + symmetrize_c4v ---
 chi   E_general       E_c4v     rel     chi   E_general        E_c4v      rel
   8  −0.659731739  −0.682908611  3.5e-02    8  −0.659843578478  −0.659843578478  0.0e+00
  12  −0.659731740  −0.682918654  3.5e-02   12  −0.659843578478  −0.659843578478  1.7e-16
  16  −0.659731740  −0.682914810  3.5e-02   16  −0.659843578478  −0.659843578478  3.4e-16
  24  −0.659731740  −0.682925748  3.5e-02   24  −0.659843578478  −0.659843578478  5.0e-16
```

**The projected arm agrees to machine precision, flat in χ** — `5.0e-16`. That is the arm that tests this issue, and it reproduces the "D=2 is clean" claim on a physical state. It is a *tighter* result than the random-fixture sweep in the issue body, which had the seed-17 outlier at `1.5e-08`.

## Two methodological notes for anyone repeating this

**1. The optimizer leaves the C4v manifold, so the as-is arm is uninformative.** From a start that is C4v to `1.4e-16`, 200 L-BFGS steps drift to `1.8e-01`. The `3.5e-02` gap in the left-hand table is therefore the expected artifact of running a C4v-assuming CTM on a non-C4v state — it is *not* evidence for this issue, and I'm not reporting it as such. Only the projected arm counts. Constraining the optimization to the symmetric manifold (rather than projecting afterwards) would remove the need for that caveat.

**2. Projection improves the state here; it does not degrade it.** The projected energy `−0.659843578478` is *below* the raw `−0.659731739524`, cutting the error against the reference from `3.17e-04` to `1.48e-04`. The D=2 ground state is C4v and the optimizer simply hadn't finished symmetrizing, so `symmetrize_c4v` moves toward the physical state rather than away from it. Worth knowing before treating the projected state as a compromise.

## A failed first attempt, recorded so it isn't repeated

60 Adam steps at `lr=1e-2` from a random (non-symmetric) start reached `E=−0.0979` against the converged `−0.6599`, with C4v deviation `0.87`. It printed large gaps at both D=2 and D=3 that looked like hits — `7.97e-01` and `2.57e-01` as-is, `5.6e-02` and `4.7e-03` projected. **All four are artifacts of an unconverged, non-symmetric state**, which is exactly what the G1 gate now catches. Had the D=2 control not been run alongside, that run would have read as a strong confirmation of this issue at D=3. Any version of this check needs the D=2 control and a convergence gate, or it produces confident nonsense.

## Status

D=3 is running with the identical recipe (200 L-BFGS steps, χ_opt=12). There's no published D=3 reference to gate against, so I'll apply the necessary condition instead — `E_opt` must land below the D=2 optimum — and discard the run if it doesn't. Expect ≈ `−0.664`.
